src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -119,12 +148,35 @@ async Task EnsureLong()
 
                 try
                 {
+                    var timeoutCt = CancellationToken.None;
+                    if (async)
+                    {
+                        if (ReadTimeout > 0)
+                        {
+                            // We reuse the timeout's cancellation token source as long as it hasn't fired, but once it has
+                            // there's no way to reset it (see https://github.com/dotnet/runtime/issues/4694)
+                            var timeoutTimeSpan = TimeSpan.FromMilliseconds(ReadTimeout);
+                            _timeoutCts.CancelAfter(timeoutTimeSpan);
+                            if (_timeoutCts.IsCancellationRequested)
+                            {
+                                _timeoutCts.Dispose();
+                                _timeoutCts = new CancellationTokenSource(timeoutTimeSpan);
+                            }
+                            timeoutCt = _timeoutCts.Token;
```

Comment:
This looks good, but it's missing [the CancelAfter call](https://github.com/npgsql/npgsql/blob/4bd4115d9605ddf12d63b927def67c8812cf74c8/src/Npgsql/ConnectorPool.Multiplexing.cs#L219) after the operation completes, to disable the timeout. This is necessary to make sure the cancellation token source doesn't needlessly expire all the time.